libs/tools/barista/src/generators/ds-navigation.ts

```diff
@@ -0,0 +1,97 @@
+/**
+ * @license
+ * Copyright 2020 Dynatrace LLC
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { join, basename } from 'path';
+import {
+  promises as fs,
+  readdirSync,
+  lstatSync,
+} from 'fs';
+
+import {
+  DsSideNavContent,
+} from '../types';
+
+const distDir = './dist/next-data';
+
+
+function addNavSection(sideNav: DsSideNavContent, navGroup: string) {
+  let newSideNav = {...sideNav};
+  for (const section of newSideNav.sections) {
+    if(section.title && navGroup == section.title){
+      return newSideNav;
+    }
+  }
+
+  newSideNav.sections.push({title: navGroup, items: []});
+  return newSideNav;
+}
+
+/** Check if given path is a directory within distDir. */
+function isDirectory(path: string): boolean {
+  try {
+    return lstatSync(join(distDir, path)).isDirectory();
+  } catch {
+    return false;
+  }
+}
+
+/** Builds navigation */
+export const navigationBuilder = async () => {
+  const allDirectories = readdirSync(distDir).filter((dirPath) =>
+    isDirectory(dirPath),
+  );
+
+  for (const directory of allDirectories) {
+    const currentPath = join(distDir, directory);
+    const files = readdirSync(currentPath);
+    const fileMap = new Map<string, any>();
+
+    let sideNav: DsSideNavContent = {
+      sections: [],
+    };
+
+    for (const file of files) {
+      const fileContent = await fs.readFile(join(currentPath, file), { encoding: 'utf-8' });
```

Comment:
in this case the sync version would be way faster then the async version if you want to perform it async then do something like

```typescript
const fileContents = await Promise.all(files.map(file => fs.readFile(join(currentPath, file), { encoding: 'utf-8' }));
fileContents.forEach(fileContent => {
  // do the stuff
})
```